crates/ruff/src/rules/flake8_todo/rules.rs

```diff
@@ -0,0 +1,525 @@
+                }
+            }
+
+            diagnostics.push(Diagnostic::new(MissingLinkInTodo, tag.range));
+        } else {
+            // There's a TODO on the last line of the file, so there can't be a link after it.
+            diagnostics.push(Diagnostic::new(MissingLinkInTodo, tag.range));
+        }
+    }
+
+    diagnostics
+}
+
+/// Returns the tag pulled out of a given comment if it exists.
+fn detect_tag<'a>(comment: &'a str, comment_range: &'a TextRange) -> Option<Tag<'a>> {
+    let Some(regex_index) = TODO_REGEX_SET.matches(comment).into_iter().next() else {
+        return None;
+    };
+
+    let tag_length = *PATTERN_TAG_LENGTH.get(&regex_index).unwrap();
+
+    let mut tag_start_offset = 0usize;
+    for (i, char) in comment.chars().enumerate() {
+        // Regex ensures that the first letter in the comment is the first letter of the tag.
+        if char.is_alphabetic() {
+            tag_start_offset = i;
+            break;
+        }
+    }
+
+    Some(Tag {
+        content: comment
+            .get(tag_start_offset..tag_start_offset + tag_length)
+            .unwrap(),
+        range: TextRange::at(
+            comment_range.start() + TextSize::try_from(tag_start_offset).ok().unwrap(),
+            TextSize::try_from(tag_length).ok().unwrap(),
+        ),
+    })
+}
+
+/// Check that the tag is valid. This function modifies `diagnostics` in-place.
+fn check_for_tag_errors(
+    tag: &Tag,
+    diagnostics: &mut Vec<Diagnostic>,
+    autofix: flags::Autofix,
+    settings: &Settings,
+) {
+    if tag.content != "TODO" {
+        // TDO-001
+        diagnostics.push(Diagnostic::new(
+            InvalidTodoTag {
+                tag: tag.content.to_string(),
+            },
+            tag.range,
+        ));
+
+        if tag.content.to_uppercase() == "TODO" {
+            // TDO-006
+            let mut invalid_capitalization = Diagnostic::new(
+                InvalidCapitalizationInTodo {
+                    tag: tag.content.to_string(),
+                },
+                tag.range,
+            );
+
+            if autofix.into() && settings.rules.should_fix(Rule::InvalidCapitalizationInTodo) {
+                invalid_capitalization.set_fix(Fix::unspecified(Edit::replacement(
+                    "TODO".to_string(),
+                    tag.range.start(),
+                    tag.range.end(),
+                )));
+            }
+
+            diagnostics.push(invalid_capitalization);
+        }
+    }
+}
+
+/// Checks for "static" errors in the comment - missing colon, missing author, etc. This function
+/// modifies `diagnostics` in-place.
+fn check_for_static_errors(
+    comment: &str,
+    comment_range: TextRange,
+    tag: &Tag,
+    diagnostics: &mut Vec<Diagnostic>,
+) {
+    // Relative offset of the current character from the start of the comment.
+    let mut relative_offset: usize = usize::try_from(tag.range.end() - comment_range.start())
+        .ok()
+        .unwrap();
```

Comment:
So it's technically the distance of the `#` denoting the start of the comment and the last character of the tag. So if you had: `# TODO`, `tag.range.len()` would be 4 but `relative_offset` would be 6. This is because `flake8_todo` matches [all whitepsace](https://github.com/orsinium-labs/flake8-todos/blob/master/flake8_todos/_rules.py#L12) before the tag. So something like `#    TODO` would still match our regex, where `relative offset` would be 8, while `tag.range.len()` would still be 4